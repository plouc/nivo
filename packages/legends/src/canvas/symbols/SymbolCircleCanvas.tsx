import { SymbolProps } from '../../types'

export const renderSymbolCircleToCanvas = (
    ctx: CanvasRenderingContext2D,
    {
        x,
        y,
        size,
        fill,
    }: //opacity = 1,
    //borderWidth = 0,
    //borderColor = 'transparent',
    SymbolProps
) => {
    //console.log('render circle ' + [x, y, size, fill])
    ctx.beginPath()
    ctx.arc(x, y, size / 2, 0, 2 * Math.PI)
    ctx.fillStyle = fill ?? 'black'
    ctx.fill()
}
