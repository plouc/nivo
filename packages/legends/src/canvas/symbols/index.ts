export * from './SymbolCircleCanvas'
export * from './SymbolSquareCanvas'
export * from './SymbolDiamondCanvas'
export * from './SymbolTriangleCanvas'
export * from './SymbolInvertedTriangleCanvas'
