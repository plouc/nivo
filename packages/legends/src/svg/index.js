/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
export { default as BoxLegendSvg } from './BoxLegendSvg'
export { default as LegendSvg } from './LegendSvg'
export { default as LegendSvgItem } from './LegendSvgItem'
