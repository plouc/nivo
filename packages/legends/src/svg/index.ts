export * from './BoxLegendSvg'
export * from './LegendSvg'
export * from './LegendSvgItem'
