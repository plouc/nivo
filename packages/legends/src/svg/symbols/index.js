/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
export { default as SymbolCircle } from './SymbolCircle'
export { default as SymbolDiamond } from './SymbolDiamond'
export { default as SymbolSquare } from './SymbolSquare'
export { default as SymbolTriangle } from './SymbolTriangle'
