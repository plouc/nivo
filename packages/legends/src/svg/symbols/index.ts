/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import { FunctionComponent } from 'react'
import { LegendDatum } from '../../props'
import { SymbolCircle } from './SymbolCircle'
import { SymbolDiamond } from './SymbolDiamond'
import { SymbolSquare } from './SymbolSquare'
import { SymbolTriangle } from './SymbolTriangle'

export * from './SymbolCircle'
export * from './SymbolDiamond'
export * from './SymbolSquare'
export * from './SymbolTriangle'

export const symbolByShape = {
    circle: SymbolCircle,
    diamond: SymbolDiamond,
    square: SymbolSquare,
    triangle: SymbolTriangle,
}

export interface LegendSymbolProps {
    id: LegendDatum['id']
    x: number
    y: number
    size: number
    fill: string
    borderWidth: number
    borderColor: string
}

export type LegendCustomSymbol = FunctionComponent<LegendSymbolProps>

export type LegendSymbolShape = keyof typeof symbolByShape | LegendCustomSymbol
