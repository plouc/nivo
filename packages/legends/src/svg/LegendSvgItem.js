/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import React, { useState, useCallback } from 'react'
import PropTypes from 'prop-types'
import isFunction from 'lodash/isFunction'
import { useTheme } from '@nivo/core'
import { datumPropType, symbolPropTypes, interactivityPropTypes } from '../props'
import { computeItemLayout } from '../compute'
import { SymbolCircle, SymbolDiamond, SymbolSquare, SymbolTriangle } from './symbols'

const symbolByShape = {
    circle: SymbolCircle,
    diamond: SymbolDiamond,
    square: SymbolSquare,
    triangle: SymbolTriangle,
}

const LegendSvgItem = ({
    x,
    y,
    width,
    height,
    data,
    direction,
    justify,
    textColor,
    background,
    opacity,

    symbolShape,
    symbolSize,
    symbolSpacing,
    symbolBorderWidth,
    symbolBorderColor,

    onClick,
    onMouseEnter,
    onMouseLeave,

    effects,
}) => {
    const [style, setStyle] = useState({})
    const theme = useTheme()

    const handleClick = useCallback(event => onClick && onClick(data, event), [onClick, data])
    const handleMouseEnter = useCallback(
        event => {
            if (effects.length > 0) {
                const applyEffects = effects.filter(({ on }) => on === 'hover')
                const style = applyEffects.reduce(
                    (acc, effect) => ({
                        ...acc,
                        ...effect.style,
                    }),
                    {}
                )
                setStyle(style)
            }

            if (onMouseEnter === undefined) return
            onMouseEnter(data, event)
        },
        [onMouseEnter, data, effects]
    )
    const handleMouseLeave = useCallback(() => {
        if (effects.length > 0) {
            const applyEffects = effects.filter(({ on }) => on !== 'hover')
            const style = applyEffects.reduce(
                (acc, effect) => ({
                    ...acc,
                    ...effect.style,
                }),
                {}
            )
            setStyle(style)
        }

        if (onMouseLeave === undefined) return
        onMouseLeave(data, event)
    }, [onMouseLeave, data, effects])

    const { symbolX, symbolY, labelX, labelY, labelAnchor, labelAlignment } = computeItemLayout({
        direction,
        justify,
        symbolSize: style.symbolSize || symbolSize,
        symbolSpacing,
        width,
        height,
    })

    const isInteractive = [onClick, onMouseEnter, onMouseLeave].some(
        handler => handler !== undefined
    )

    let Symbol
    if (isFunction(symbolShape)) {
        Symbol = symbolShape
    } else {
        Symbol = symbolByShape[symbolShape]
    }

    return (
        <g
            transform={`translate(${x},${y})`}
            style={{
                opacity: style.itemOpacity !== undefined ? style.itemOpacity : opacity,
            }}
        >
            <rect
                width={width}
                height={height}
                fill={style.itemBackground || background}
                style={{
                    cursor: isInteractive ? 'pointer' : 'auto',
                }}
                onClick={handleClick}
                onMouseEnter={handleMouseEnter}
                onMouseLeave={handleMouseLeave}
            />
            {React.createElement(Symbol, {
                id: data.id,
                x: symbolX,
                y: symbolY,
                size: style.symbolSize || symbolSize,
                fill: data.fill || data.color,
                borderWidth:
                    style.symbolBorderWidth !== undefined
                        ? style.symbolBorderWidth
                        : symbolBorderWidth,
                borderColor: style.symbolBorderColor || symbolBorderColor,
            })}
            <text
                textAnchor={labelAnchor}
                style={{
                    ...theme.legends.text,
                    fill: style.itemTextColor || theme.legends.text.fill || textColor,
                    dominantBaseline: labelAlignment,
                    pointerEvents: 'none',
                    userSelect: 'none',
                }}
                x={labelX}
                y={labelY}
            >
                {data.label}
            </text>
        </g>
    )
}

LegendSvgItem.displayName = 'LegendSvgItem'
LegendSvgItem.propTypes = {
    data: datumPropType.isRequired,

    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,

    textColor: PropTypes.string,
    background: PropTypes.string,
    opacity: PropTypes.number,

    direction: PropTypes.oneOf(['left-to-right', 'right-to-left', 'top-to-bottom', 'bottom-to-top'])
        .isRequired,
    justify: PropTypes.bool.isRequired,

    ...symbolPropTypes,
    ...interactivityPropTypes,
}
LegendSvgItem.defaultProps = {
    direction: 'left-to-right',
    justify: false,

    textColor: 'black',
    background: 'transparent',
    opacity: 1,

    symbolShape: 'square',
    symbolSize: 16,
    symbolSpacing: 8,
    symbolBorderWidth: 0,
    symbolBorderColor: 'transparent',

    effects: [],
}

export default LegendSvgItem
