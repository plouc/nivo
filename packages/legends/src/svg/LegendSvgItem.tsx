import { useState, useCallback } from 'react'
import * as React from 'react'
import { useTheme } from '@nivo/core'
import { LegendSvgItemProps } from '../types'
import { computeItemLayout } from '../compute'
import { SymbolCircle, SymbolDiamond, SymbolSquare, SymbolTriangle } from './symbols'

type Style = Partial<{
    itemBackground: string
    itemOpacity: number
    itemTextColor: string
    symbolBorderColor: string
    symbolBorderWidth: number
    symbolSize: number
}>

const symbolByShape = {
    circle: SymbolCircle,
    diamond: SymbolDiamond,
    square: SymbolSquare,
    triangle: SymbolTriangle,
}

export const LegendSvgItem = ({
    x,
    y,
    width,
    height,
    data,
    direction = 'left-to-right',
    justify = false,
    textColor,
    background = 'transparent',
    opacity = 1,

    symbolShape = 'square',
    symbolSize = 16,
    symbolSpacing = 8,
    symbolBorderWidth = 0,
    symbolBorderColor = 'transparent',

    onClick,
    onMouseEnter,
    onMouseLeave,
    toggleSerie,

    effects,
}: LegendSvgItemProps) => {
    const [style, setStyle] = useState<Style>({})
    const theme = useTheme()

    const handleMouseEnter = useCallback(
        (event: React.MouseEvent<SVGRectElement>) => {
            if (effects) {
                const applyEffects = effects.filter(({ on }) => on === 'hover')
                const style = applyEffects.reduce(
                    (acc, effect) => ({
                        ...acc,
                        ...effect.style,
                    }),
                    {}
                )
                setStyle(style)
            }

            onMouseEnter?.(data, event)
        },
        [onMouseEnter, data, effects]
    )
    const handleMouseLeave = useCallback(
        (event: React.MouseEvent<SVGRectElement>) => {
            if (effects) {
                const applyEffects = effects.filter(({ on }) => on !== 'hover')
                const style = applyEffects.reduce(
                    (acc, effect) => ({
                        ...acc,
                        ...effect.style,
                    }),
                    {}
                )
                setStyle(style)
            }

            onMouseLeave?.(data, event)
        },
        [onMouseLeave, data, effects]
    )

    const { symbolX, symbolY, labelX, labelY, labelAnchor, labelAlignment } = computeItemLayout({
        direction,
        justify,
        symbolSize: style.symbolSize ?? symbolSize,
        symbolSpacing,
        width,
        height,
    })

    const isInteractive = [onClick, onMouseEnter, onMouseLeave, toggleSerie].some(
        handler => handler !== undefined
    )

    const SymbolShape = typeof symbolShape === 'function' ? symbolShape : symbolByShape[symbolShape]

    return (
        <g
            transform={`translate(${x},${y})`}
            style={{
                opacity: style.itemOpacity ?? opacity,
            }}
        >
            <rect
                width={width}
                height={height}
                fill={style.itemBackground ?? background}
                style={{
                    cursor: isInteractive ? 'pointer' : 'auto',
                }}
                onClick={event => {
                    onClick?.(data, event)
                    toggleSerie?.(data.id)
                }}
                onMouseEnter={handleMouseEnter}
                onMouseLeave={handleMouseLeave}
            />
            {React.createElement(SymbolShape, {
                id: data.id,
                x: symbolX,
                y: symbolY,
                size: style.symbolSize ?? symbolSize,
                fill: data.fill ?? data.color ?? 'black',
                borderWidth: style.symbolBorderWidth ?? symbolBorderWidth,
                borderColor: style.symbolBorderColor ?? symbolBorderColor,
                ...(data.hidden ? theme.legends.hidden.symbol : undefined),
            })}
            <text
                textAnchor={labelAnchor}
                style={{
                    ...theme.legends.text,
                    fill: style.itemTextColor ?? textColor ?? theme.legends.text.fill ?? 'black',
                    dominantBaseline: labelAlignment,
                    pointerEvents: 'none',
                    userSelect: 'none',
                    ...(data.hidden ? theme.legends.hidden.text : undefined),
                }}
                x={labelX}
                y={labelY}
            >
                {data.label}
            </text>
        </g>
    )
}
