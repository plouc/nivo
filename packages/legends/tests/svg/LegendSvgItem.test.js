/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, RaphaÃ«l Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import { defaultTheme } from '@nivo/core'
import LegendSvgItem from '../../src/svg/LegendSvgItem'
import * as shapes from '../../src/svg/symbols'
import {
    DIRECTION_LEFT_TO_RIGHT,
    DIRECTION_RIGHT_TO_LEFT,
    DIRECTION_TOP_TO_BOTTOM,
    DIRECTION_BOTTOM_TO_TOP,
} from '../../src/constants'

const commonProps = {
    x: 0,
    y: 0,
    width: 200,
    height: 36,
    theme: defaultTheme,
    data: {
        id: 'testing',
        label: 'testing',
        color: 'red',
    },
}

const directions = [
    DIRECTION_LEFT_TO_RIGHT,
    DIRECTION_RIGHT_TO_LEFT,
    DIRECTION_TOP_TO_BOTTOM,
    DIRECTION_BOTTOM_TO_TOP,
]

directions.forEach(direction => {
    it(`should support ${direction} direction`, () => {
        const component = renderer.create(<LegendSvgItem {...commonProps} direction={direction} />)

        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })

    it(`should support ${direction} direction justified`, () => {
        const component = renderer.create(
            <LegendSvgItem {...commonProps} direction={direction} justify={true} />
        )

        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
})

describe('symbolShape', () => {
    it('should support circle shape', () => {
        const wrapper = mount(
            <svg>
                <LegendSvgItem {...commonProps} symbolShape="circle" />
            </svg>
        )

        expect(wrapper.find(shapes.SymbolCircle)).toHaveLength(1)
    })

    it('should support diamond shape', () => {
        const wrapper = mount(
            <svg>
                <LegendSvgItem {...commonProps} symbolShape="diamond" />
            </svg>
        )

        expect(wrapper.find(shapes.SymbolDiamond)).toHaveLength(1)
    })

    it('should support square shape', () => {
        const wrapper = mount(
            <svg>
                <LegendSvgItem {...commonProps} symbolShape="square" />
            </svg>
        )

        expect(wrapper.find(shapes.SymbolSquare)).toHaveLength(1)
    })

    it('should support triangle shape', () => {
        const wrapper = mount(
            <svg>
                <LegendSvgItem {...commonProps} symbolShape="triangle" />
            </svg>
        )

        expect(wrapper.find(shapes.SymbolTriangle)).toHaveLength(1)
    })

    it('should support custom shape', () => {
        const CustomShape = () => <g />
        const wrapper = mount(
            <svg>
                <LegendSvgItem {...commonProps} symbolShape={CustomShape} />
            </svg>
        )

        expect(wrapper.find(CustomShape)).toHaveLength(1)
    })
})
