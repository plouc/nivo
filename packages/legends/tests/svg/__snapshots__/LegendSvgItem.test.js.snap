// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should support bottom-to-top direction 1`] = `
<g
  style={
    Object {
      "opacity": 1,
    }
  }
  transform="translate(0,0)"
>
  <rect
    fill="transparent"
    height={36}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "auto",
      }
    }
    width={200}
  />
  <rect
    fill="red"
    height={16}
    stroke="transparent"
    strokeWidth={0}
    style={
      Object {
        "pointerEvents": "none",
      }
    }
    width={16}
    x={92}
    y={20}
  />
  <text
    style={
      Object {
        "dominantBaseline": "alphabetic",
        "fill": "#333333",
        "fontFamily": "sans-serif",
        "fontSize": 11,
        "pointerEvents": "none",
        "userSelect": "none",
      }
    }
    textAnchor="middle"
    x={100}
    y={12}
  >
    testing
  </text>
</g>
`;

exports[`should support bottom-to-top direction justified 1`] = `
<g
  style={
    Object {
      "opacity": 1,
    }
  }
  transform="translate(0,0)"
>
  <rect
    fill="transparent"
    height={36}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "auto",
      }
    }
    width={200}
  />
  <rect
    fill="red"
    height={16}
    stroke="transparent"
    strokeWidth={0}
    style={
      Object {
        "pointerEvents": "none",
      }
    }
    width={16}
    x={92}
    y={20}
  />
  <text
    style={
      Object {
        "dominantBaseline": "text-before-edge",
        "fill": "#333333",
        "fontFamily": "sans-serif",
        "fontSize": 11,
        "pointerEvents": "none",
        "userSelect": "none",
      }
    }
    textAnchor="middle"
    x={100}
    y={0}
  >
    testing
  </text>
</g>
`;

exports[`should support left-to-right direction 1`] = `
<g
  style={
    Object {
      "opacity": 1,
    }
  }
  transform="translate(0,0)"
>
  <rect
    fill="transparent"
    height={36}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "auto",
      }
    }
    width={200}
  />
  <rect
    fill="red"
    height={16}
    stroke="transparent"
    strokeWidth={0}
    style={
      Object {
        "pointerEvents": "none",
      }
    }
    width={16}
    x={0}
    y={10}
  />
  <text
    style={
      Object {
        "dominantBaseline": "central",
        "fill": "#333333",
        "fontFamily": "sans-serif",
        "fontSize": 11,
        "pointerEvents": "none",
        "userSelect": "none",
      }
    }
    textAnchor="start"
    x={24}
    y={18}
  >
    testing
  </text>
</g>
`;

exports[`should support left-to-right direction justified 1`] = `
<g
  style={
    Object {
      "opacity": 1,
    }
  }
  transform="translate(0,0)"
>
  <rect
    fill="transparent"
    height={36}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "auto",
      }
    }
    width={200}
  />
  <rect
    fill="red"
    height={16}
    stroke="transparent"
    strokeWidth={0}
    style={
      Object {
        "pointerEvents": "none",
      }
    }
    width={16}
    x={0}
    y={10}
  />
  <text
    style={
      Object {
        "dominantBaseline": "central",
        "fill": "#333333",
        "fontFamily": "sans-serif",
        "fontSize": 11,
        "pointerEvents": "none",
        "userSelect": "none",
      }
    }
    textAnchor="end"
    x={200}
    y={18}
  >
    testing
  </text>
</g>
`;

exports[`should support right-to-left direction 1`] = `
<g
  style={
    Object {
      "opacity": 1,
    }
  }
  transform="translate(0,0)"
>
  <rect
    fill="transparent"
    height={36}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "auto",
      }
    }
    width={200}
  />
  <rect
    fill="red"
    height={16}
    stroke="transparent"
    strokeWidth={0}
    style={
      Object {
        "pointerEvents": "none",
      }
    }
    width={16}
    x={184}
    y={10}
  />
  <text
    style={
      Object {
        "dominantBaseline": "central",
        "fill": "#333333",
        "fontFamily": "sans-serif",
        "fontSize": 11,
        "pointerEvents": "none",
        "userSelect": "none",
      }
    }
    textAnchor="end"
    x={176}
    y={18}
  >
    testing
  </text>
</g>
`;

exports[`should support right-to-left direction justified 1`] = `
<g
  style={
    Object {
      "opacity": 1,
    }
  }
  transform="translate(0,0)"
>
  <rect
    fill="transparent"
    height={36}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "auto",
      }
    }
    width={200}
  />
  <rect
    fill="red"
    height={16}
    stroke="transparent"
    strokeWidth={0}
    style={
      Object {
        "pointerEvents": "none",
      }
    }
    width={16}
    x={184}
    y={10}
  />
  <text
    style={
      Object {
        "dominantBaseline": "central",
        "fill": "#333333",
        "fontFamily": "sans-serif",
        "fontSize": 11,
        "pointerEvents": "none",
        "userSelect": "none",
      }
    }
    textAnchor="start"
    x={0}
    y={18}
  >
    testing
  </text>
</g>
`;

exports[`should support top-to-bottom direction 1`] = `
<g
  style={
    Object {
      "opacity": 1,
    }
  }
  transform="translate(0,0)"
>
  <rect
    fill="transparent"
    height={36}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "auto",
      }
    }
    width={200}
  />
  <rect
    fill="red"
    height={16}
    stroke="transparent"
    strokeWidth={0}
    style={
      Object {
        "pointerEvents": "none",
      }
    }
    width={16}
    x={92}
    y={0}
  />
  <text
    style={
      Object {
        "dominantBaseline": "text-before-edge",
        "fill": "#333333",
        "fontFamily": "sans-serif",
        "fontSize": 11,
        "pointerEvents": "none",
        "userSelect": "none",
      }
    }
    textAnchor="middle"
    x={100}
    y={24}
  >
    testing
  </text>
</g>
`;

exports[`should support top-to-bottom direction justified 1`] = `
<g
  style={
    Object {
      "opacity": 1,
    }
  }
  transform="translate(0,0)"
>
  <rect
    fill="transparent"
    height={36}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "cursor": "auto",
      }
    }
    width={200}
  />
  <rect
    fill="red"
    height={16}
    stroke="transparent"
    strokeWidth={0}
    style={
      Object {
        "pointerEvents": "none",
      }
    }
    width={16}
    x={92}
    y={0}
  />
  <text
    style={
      Object {
        "dominantBaseline": "alphabetic",
        "fill": "#333333",
        "fontFamily": "sans-serif",
        "fontSize": 11,
        "pointerEvents": "none",
        "userSelect": "none",
      }
    }
    textAnchor="middle"
    x={100}
    y={36}
  >
    testing
  </text>
</g>
`;
